### PR TITLE
Show gRPCs in requests dialog

### LINKED ISSUE #3271: gRPC requests do not appear in request navigation (Quick Switch, Recent)
**Describe the bug**

Invoking the Quick Switch dialog or the Recent Requests dialog do not list any gRPC requests.

**To Reproduce**
1. Create a new collection "gRPC Bug Repro"
2. Add a gRPC request, add a proto file with a copy of `grpcbin.proto` from https://grpcb.in/
3. Duplicate the request, choose a different RPC like `/GRPCBin/Empty`
4. Invoke keyboard shortcut for "Switch Requests", see nothing in dialog
5. Invoke keyboard shortcut for "Show Recent Requests", see nothing in dialog
6. Add a new HTTP request to the collection
7. Invoke the same shortcuts, see the HTTP request present

**Expected behavior**
Request listings for gRPC just as I get with requests using HTTP, GraphQL, etc.

**Desktop (please complete the following information):**
 - OS: macOS 11.2.3
 - Installation Method: homebrew
 - App Version 2021.2.2

### PR DESCRIPTION
**Pull request by vincendep**
changelog(Improvements): gRPC requests will now appear in the request switcher

 Closes #3271

### CODE REVIEW

**Review by develohpanda — CHANGES_REQUESTED** [review]
Thanks for the PR! We do need to add some UI for GRPC requests in the request switcher as well. For instance, the following is what I see when hitting `cmd + p`.

Ideally we follow a similar pattern to REST requests.

image

I also noticed the following lines: https://github.com/vincendep/insomnia/blob/21047c818165a63c79d9484edf723e39cd85416b/packages/insomnia-app/app/ui/components/modals/request-switcher-modal.tsx#L34-L35

We'll need to remove this (now that GRPC requests have been introduced to the request switcher), and determine whether there are any implications. I couldn't see any while browsing this file.

Lastly, could you also please (at the very least) add screenshots or gifs to PR descriptions? It really helps with code reviews. Thank you! 🙏🏽

**Comment by vincendep** [comment]
Hi @develohpanda, so sorry for my lazyness 😄, I'll be more explicative for my next pull requests. Speaking of UI changes, do you think would be ok to show only "gRPC" (as Add Request dropdown), 

Schermata 2021-10-18 alle 10 44 58

or maybe we should include the selected gRPC  method some how?

**Comment by develohpanda** [comment]
No worries at all! It would be nice to show the url + selected grpc method name. Should be fairly straightforward to get that information I think

**Comment by filfreire** [comment]
@vincendep, in case you are interested you could achieve what @develohpanda mentioned with something like:

Near line 224:
```diff
       finalUrl = joinUrlAndQueryString(finalUrl, buildQueryStringFromParams(request.parameters));
     }
 
+    if (isGrpcRequest(request)) {
+      finalUrl = request.url + request.protoMethodName;
+    }
+
     const match = fuzzyMatchAll(
       searchStrings,
       [request.name, finalUrl, request.method || '', this._groupOf(request).join('/')],
```

Near line 477:
```diff
                       </div>
                       <div className="margin-left-xs faint">
                         { isRequest(r) ? <MethodTag method={r.method} /> : null }
-                        <Highlight search={searchString} text={r.url} />
+                        <Highlight
+                          search={searchString}
+                          text={ isGrpcRequest(r) ? r.url + r.protoMethodName : r.url }
+                        />
                       </div>
                     </Button>
                   </li>

```


Example of how it would look like:

Screenshot 2021-12-13 at 12 06 53
Screenshot 2021-12-13 at 12 02 32

**Comment by vincendep** [comment]
Sorry about end of communications by I didn't found time to work on this.  
Hi @filfreire, thanks for your help, I'll give it a try soon and see if I can close this. Btw I think that would be nice to show also the gRPC tag in place of the method tag

**Comment by vincendep** [comment]
Here a screenshoot of the update

Schermata 2021-12-18 alle 10 06 01

**Review by filfreire — APPROVED** [review]
Looks good to me, great work @vincendep 👍 

Screenshot 2022-01-04 at 17 32 35

**Comment by filfreire** [comment]
FYI @dimitropoulos @develohpanda this PR would be a good candidate to get merged before the next release.

File: packages/insomnia-app/app/ui/components/modals/request-switcher-modal.tsx [hunk 1]
```diff
@@ -243,17 +248,20 @@ class RequestSwitcherModal extends PureComponent<Props, State> {
...
       lastActiveMap[meta.parentId] = meta.lastActive;
     }
+    for (const meta of grpcRequestMetas) {
+      lastActiveMap[meta.parentId] = meta.lastActive;
+    }
 
     // OPTIMIZATION: This only filters if we have a filter
     let matchedRequests = workspaceChildren
-      .filter(isRequest)
+      .filter(child => isRequest(child) || isGrpcRequest(child))
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/components/modals/request-switcher-modal.tsx]: Unfortunately with this change we lose type safety (`matchedRequest` should not be of type `RequestGroup`). There's a known bug in TS regarding this, and I haven't seen any reliable workarounds for it.

After
image

Before
image

I've updated some of the type definitions here to bring the `as` cast alongside: e3486969243ca51f81b6788ab24692f5dea6b2a8

**Review by develohpanda — APPROVED** [review]
Thanks for the updates @vincendep! I pushed a few small commits, please take a look!

These commits remove a redundant comment and improves some TS types that were incorrect after the changes in this PR (due to a quirk of TS and not your fault 🤗)

File: packages/insomnia-app/app/ui/components/modals/request-switcher-modal.tsx (hunk 1 above)
Comment (vincendep) [reply packages/insomnia-app/app/ui/components/modals/request-switcher-modal.tsx]: Hi @develohpanda. Thanks for the explanation, given thath I'm a beginner with TS every advice is welcome

**Review by dimitropoulos — APPROVED** [review]

**Pull request merged**